fish-rust/src/builtins/string/split.rs

```diff
@@ -0,0 +1,296 @@
+                        return Err(StringError::InvalidArgs("Invalid range value for field"))
+                    }
+                };
+            }
+            'a' => self.allow_empty = true,
+            _ => return Err(StringError::UnknownOption),
+        }
+        return Ok(());
+    }
+
+    fn take_args(
+        &mut self,
+        optind: &mut usize,
+        args: &[&'args wstr],
+        streams: &mut io_streams_t,
+    ) -> Option<c_int> {
+        if self.is_split0 {
+            return STATUS_CMD_OK;
+        }
+        let Some(arg) = args.get(*optind).copied() else {
+            string_error!(streams, BUILTIN_ERR_ARG_COUNT0, args[0]);
+            return STATUS_INVALID_ARGS;
+        };
+        *optind += 1;
+        self.sep = arg;
+        return STATUS_CMD_OK;
+    }
+
+    fn handle(
+        &mut self,
+        _parser: &mut parser_t,
+        streams: &mut io_streams_t,
+        optind: &mut usize,
+        args: &[&'args wstr],
+    ) -> Option<c_int> {
+        if self.fields.is_empty() && self.allow_empty {
+            streams.err.append(wgettext_fmt!(
+                BUILTIN_ERR_COMBO2,
+                args[0],
+                "--allow-empty is only valid with --fields"
+            ));
+            return STATUS_INVALID_ARGS;
+        }
+
+        let sep = &self.sep;
+        let mut all_splits: Vec<Vec<Cow<'args, wstr>>> = Vec::new();
+        let mut split_count = 0usize;
+        let mut arg_count = 0usize;
+
+        for (arg, _) in Arguments::new(args, optind, streams, !self.is_split0) {
+            let splits: Vec<Cow<'args, wstr>> = match (self.split_from, arg) {
+                (Direction::Right, arg) => {
+                    let mut rev = arg.into_owned();
+                    rev.as_char_slice_mut().reverse();
+                    split_about(&rev, sep, self.max, self.no_empty)
+                        .into_iter()
+                        // If we are from the right, split_about gave us reversed strings, in reversed order!
+                        .map(|s| Cow::Owned(s.chars().rev().collect::<WString>()))
+                        .rev()
+                        .collect()
+                }
+                // we need to special-case the Cow::Borrowed case, since
+                // let arg: &'args wstr = &arg;
+                // does not compile since `arg` can be dropped at the end of this scope
+                // making the reference invalid if it is owned.
+                (Direction::Left, Cow::Borrowed(arg)) => {
+                    split_about(arg, sep, self.max, self.no_empty)
+                        .into_iter()
+                        .map(Cow::Borrowed)
+                        .collect()
+                }
+                (Direction::Left, Cow::Owned(arg)) => {
+                    split_about(&arg, sep, self.max, self.no_empty)
+                        .into_iter()
+                        .map(|s| Cow::Owned(s.to_owned()))
```

Comment:
This _can_ be changed to a `Cow::Borrowed`, but we then need to store the owned `arg` _somewhere_ to extend it's lifetime (it will be dropped at the end of this loop iteration). I have not been able to find a "nice" solution to this without first collecting all args to a `Vec`, which prevents us from the early exit.

Reply:
this seems fine to me